### PR TITLE
Support for ansible collections in local ansible provisioner

### PR DESCRIPTION
**Pull request by wilsonge**
https://github.com/hashicorp/packer/issues/10666
Alternative to https://github.com/hashicorp/packer/pull/10667

Sorry I'm not a go developer (I usually sit in Python/PHP langs) so if i'm doing anything really stupid or anything please do tell me (and obviously review this as if submitted by an idiot 😆).

But this reuses some code from the ansible provisioned to use collections from ansible galaxy. It addresses some of the comments in #10667. Note as a difference from the remote provisioner I've added defaults for the role and collection paths so as to try and keep b/c the best I can with the existing code base (which has an opinionated roles path)

## What's different to the other PR
By default we don't install roles + collections globally but into the staging folder allowing them to be cleaned up fully at the end of a packer run.
By pulling in some of the code from #9903 we gain consistency with the ansible remote provisioner (including the force flag on the galaxy command)

### CODE REVIEW

**Comment by wilsonge** [comment]
So I think this will upload and run the correct commands. However it's not going to find the uploaded roles/collections from galaxy. The previous version of the code dropped them into the roles folder being uploaded from the local vm. but i guess this can end up with local files being overwritten by galaxy files. I think it's better to drop them into the local directory and load them with ansible as an extra directory. https://docs.ansible.com/ansible/latest/reference_appendices/config.html#default-roles-path and https://docs.ansible.com/ansible/latest/reference_appendices/config.html#collections-paths should do the trick for this. But I'm unsure how to set this in your packer cmd context compared to the stackoverflow answer here https://stackoverflow.com/questions/41133115/pass-env-var-to-exec-command

**Comment by wilsonge** [comment]
OK Well it's probably not the prettiest code ever. But I think this is functional and ready to review at least with my basic testing.

**Review by SwampDragons — CHANGES_REQUESTED** [review]
The code overall looks fine-- thanks for addressing my comments on the other PR. My main concern now is that moving forward we'll still have the same root problem, which is unnecessary configuration drift between the two provisioners. 

It would be cool to extract the command generation code into a common function that can be used by either provisioner. I think the best way to do this would be to add a new directory "github.com/hashicorp/provisioner/ansible/common" and extract and generalize the "executeGalaxy" function from each provisioner into that folder. 

In golang we can use inheritance to reuse sub-sets of our configuration structure. For example, we can create a new galaxy config that lives in our new "common" directory:

```golang
type GalaxyConfig struct {
	GalaxyFile string `mapstructure:"galaxy_file"`
	// The command to invoke ansible-galaxy. By default, this is
	// `ansible-galaxy`.
	GalaxyCommand string `mapstructure:"galaxy_command"`
	// Force overwriting an existing role.
	//  Adds `--force` option to `ansible-galaxy` command. By default, this is
	//  `false`.
	GalaxyForceInstall bool `mapstructure:"galaxy_force_install"`
	// The path to the directory on your local system in which to
	//   install the roles. Adds `--roles-path /path/to/your/roles` to
	//   `ansible-galaxy` command. By default, this is empty, and thus `--roles-path`
	//   option is not added to the command.
	RolesPath string `mapstructure:"roles_path"`
	// The path to the directory on your local system in which to
	//   install the collections. Adds `--collections-path /path/to/your/collections` to
	//   `ansible-galaxy` command. By default, this is empty, and thus `--collections-path`
	//   option is not added to the command.
	CollectionsPath string `mapstructure:"collections_path"`
}
```

Then embed it in the main Config struct of each provisioner (ansible and ansible-local) using

```golang
import (
	...

	ansiblecommon "github.com/hashicorp/provisioner/ansible/common"
)

type Config struct {
	common.PackerConfig `mapstructure:",squash"`
	ctx                 interpolate.Context

	...blahblah all the other config options
	
	ansiblecommon.GalaxyConfig `mapstructure:",squash"`
}
```
	
... and removing all of the fields that we pulled into the GalaxyConfig from the "parent" struct (I'm using Parent cause you have a Python history, but this isn't inheritance and thinking about it that way will get you in trouble).

As you probably just guessed, PackerConfig is also a separate struct that we embed in basically all our plugins. 

Where the defaulting is the same, you can even reuse the defaulting code by creating a Prepare() method on that GalaxyConfig and then calling the GalaxyConfig.Prepare() from inside the Prepare() method of the main Config.

Let me know if you have issues implementing this or if anything I've said isn't clear to you. I am happy to push up a refactor to your branch if you need a hand, but I figure I'll let you have at it if you'd prefer. 

Anyway, this definitely also needs docs for the new options added to https://github.com/hashicorp/packer/blob/master/website/content/docs/provisioners/ansible-local.mdx; Most of our provisioners auto-generate docs based on struct tags, but this one doesn't yet and it feels out of scope to ask you to implement that generation. I'll make a ticket to update the code generation somewhere else.

Finally, I'd really love to see tests for the executeGalaxy function once you've extracted it. Take a look at the file named provisioner_test.go in the ansible and/or ansible-local directories to get an idea for how to structure tests in golang, and if you get stuck ping me.

**Comment by wilsonge** [comment]
Docs uploaded. Will try and figure my way through the refactoring you proposed (is it really logical that the ansible module has the common code for the ansible-local provisioner?) - surely if it's going to be shared it should be in some other directory that isn't specific to either of the provisioners causing hard deps between the provisioners (understand that's kinda an implementation detail and can press on for now without an immediate answer).

Thanks for the layman terms explanations in python btw. Super appreciated!

**Comment by SwampDragons** [comment]
>is it really logical that the ansible module has the common code for the ansible-local provisioner?

You're right that it isn't ideal; there should be a higher level "common" directory that they both reference, but that's not how the Packer code is laid out, and said higher level code is undergoing a lot of refactoring at the moment so it's safer to keep the directory internal to one of the provisioners for now. Having hard dependencies between the provisioners isn't a "real" problem since they always ship together. The Packer team is in the process of splitting plugins out of the core, and when that happens would be a good time to make a common directory that they both reference, since both ansible components will live in the same plugin. I suggested putting the "common" directory inside the "ansible" plugin folder because eventually the directory structure of the standalone plugin will look like: 

```
packer-plugin-ansible // repo name
|_ ansible-local // referenced as ansible-local in packer template configs
|_ ansible-remote // referenced as ansible in packer template configs
|_common
```

But for now, sticking "common" inside the ansible subdirectory will keep it scoped in a nearly-logical way that will prevent other stuff from accidentally getting tossed into an ansible-specific common folder.

File: website/content/docs/provisioners/ansible-local.mdx [hunk 1]
```diff
@@ -193,6 +193,18 @@ Optional:
 - `galaxy_command` (string) - The command to invoke ansible-galaxy. By
   default, this is ansible-galaxy.
 
+- `galaxy_force_install` (bool) - Force overwriting an existing role or collection.
+   Adds `--force` option to `ansible-galaxy` command. By default, this is
+   `false`.
+
+- `galax_roles_path` (string) - The path to the directory on the remote system in which to
```
Comment (aslafy-z) [inline website/content/docs/provisioners/ansible-local.mdx:200]:
```suggestion
- `galaxy_roles_path` (string) - The path to the directory on the remote system in which to
```

**Comment by SwampDragons** [comment]
Oh shoot, the new ansible plugin repo got extracted out from under this PR. I'll migrate the PR when I have a chance.

**Comment by nywilken** [comment]
@SwampDragons @wilsonge the changes in this PR were ported over to the packer-plugin-ansible repo https://github.com/hashicorp/packer-plugin-ansible/pull/38

We will work on making the final requested changes to get the feature merged. Thanks for the contribution @wilsonge

**Pull request closed without merging**